wolfcrypt/src/asn.c

```diff
@@ -5564,45 +5564,25 @@ WOLFSSL_API int EccEnumToNID(int n)
 #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
 int wc_OBJ_sn2nid(const char *sn)
 {
-    const struct {
-        const char *sn;
-        int  nid;
-    } sn2nid[] = {
-        {WOLFSSL_COMMON_NAME, NID_commonName},
-        {WOLFSSL_COUNTRY_NAME, NID_countryName},
-        {WOLFSSL_LOCALITY_NAME, NID_localityName},
-        {WOLFSSL_STATE_NAME, NID_stateOrProvinceName},
-        {WOLFSSL_ORG_NAME, NID_organizationName},
-        {WOLFSSL_ORGUNIT_NAME, NID_organizationalUnitName},
-        {WOLFSSL_EMAIL_ADDR, NID_emailAddress},
-        {NULL, -1}};
-
-    int i;
-    #ifdef HAVE_ECC
-    int eccEnum;
-    #endif
-    WOLFSSL_ENTER("OBJ_sn2nid");
-    for(i=0; sn2nid[i].sn != NULL; i++) {
-        if(XSTRNCMP(sn, sn2nid[i].sn, XSTRLEN(sn2nid[i].sn)) == 0) {
-            return sn2nid[i].nid;
-        }
-    }
-    #ifdef HAVE_ECC
-    /* Nginx uses this OpenSSL string. */
-    if (XSTRNCMP(sn, "prime256v1", 10) == 0)
-        sn = "SECP256R1";
-    if (XSTRNCMP(sn, "secp384r1", 10) == 0)
-        sn = "SECP384R1";
-    /* find based on name and return NID */
-    for (i = 0; ecc_sets[i].size != 0 && ecc_sets[i].name != NULL; i++) {
-        if (XSTRNCMP(sn, ecc_sets[i].name, ECC_MAXNAME) == 0) {
-            eccEnum = ecc_sets[i].id;
-            /* Convert enum value in ecc_curve_id to OpenSSL NID */
-            return EccEnumToNID(eccEnum);
+    const WOLFSSL_ObjectInfo *obj_info = wolfssl_object_info;
```

Comment:
Where is wolfssl_object_info coming from???
The OpenSSL compatibility layer is meant to be a thin layer on top.
Don't mix layers.
Remember we are embedded a lot of the time so we want small data.
Don't make us build in data we don't have to.

Reply:
`wolfssl_object_info` is found in ssl.c. It is the struct that contains all of the NID information. When compiling with `OPENSSL_EXTRA_X509_SMALL` instead of `OPENSSL_EXTRA` then only information required for certificates in included in the struct. If `NO_CERTS` is defined then it is an empty struct. Should I revert this function to its previous version and modify `wolfSSL_OBJ_sn2nid` to use `wolfssl_object_info` instead of `wc_OBJ_sn2nid`?

Reply:
wc_OBJ_sn2nid calls wc_OBJ_sn2nid.
The data needs to be brought down from ssl.c into asn.c not data being referenced from asn.c but in ssl.c.
From what I can tell in a brief look wolfssl_object_info was only compiled with OPENSSL_EXTRA and not OPENSSL_EXTRA_X509_SMALL.
This move has greater implications. There are many other functions that need this data in ssl.c.
This really needs a separate pull request to fix.

Is there any reason we can't leave wc_OBJ_s2nid as it was?

Reply:
I changed it so that only a subset needed by `asn.c` would be compiled with `OPENSSL_EXTRA_X509_SMALL`. My reasoning for making `wc_OBJ_sn2nid` use `wolfssl_object_info` was to provide easy extensibility. The previous implementation had a very small subset of NID's hard coded in the function which was not enough for `wolfSSL_OBJ_sn2nid`. It is possible to keep `wc_OBJ_sn2nid` as it was and have `wolfSSL_OBJ_sn2nid` use `wolfssl_object_info` directly. I didn't want to introduce "duplicate" code and leave `wc_OBJ_sn2nid` "behind" in terms of correctness and general usability. I understand that we want to keep build size small which is why I stripped down `wolfssl_object_info`. Now I see that I removed `wc_OBJ_sn2nid`'s ability to recognize ECC names as well so those would also have to be included in the small `wolfssl_object_info` to maintain the same functionality.

Reply:
Revert the change.
Make a new pull request that fixes this up once this has been merged.